### ISSUE TITLE
Freedesktop files from user dont gets loaded

### ISSUE THREAD

**Issue by gsantner** [gmc-holle/xfdashboard#169]
First I though xfdashboard ignores the `Hidden` and `OnlyShowIn` attributes of `.desktop` files. XFCE whiskermenu shows/hide them correct, so everything needed is available in garcon. Then I found that it just it's just not loading files from the home's share directory, or at least always preferes `/usr/local/share/applications` over `$HOME/.local/share/applications`

![bildschirmfoto_2018-05-02_22-30-32](https://user-images.githubusercontent.com/6735650/39547667-8c124d24-4e58-11e8-8849-405542f4cc24.png)

![grafik](https://user-images.githubusercontent.com/6735650/39547797-eda4c86e-4e58-11e8-848e-26866f4f4a0c.png)

I have e.g. overwritten assistant, linguist and qdbusviewer so they don't show up.

`/usr/share/applications/assistant.desktop`
```
[Desktop Entry]
Name=Qt Assistant
Comment=Shows Qt documentation and examples
Exec=assistant
Icon=assistant
Terminal=false
Type=Application
Categories=Qt;Development;Documentation;
```

`$HOME/.local/share/applications/assistant.desktop`
```
[Desktop Entry]
Hidden=true
```

In the case where a file in the local share dir is available, xfdashboard should load that instead of the global one.

[Additional note](https://askubuntu.com/a/538590): If the env var is not set:

> $XDG_DATA_HOME defines the base directory relative to which user specific data files should be stored. If $XDG_DATA_HOME is either not set or empty, a default equal to $HOME/.local/share should be used.
> Applications that implement this specification must implement this behaviour, so if $XDG_DATA_HOME is not defined they must use $HOME/.local/share as default.

**Comment by gsantner** [comment]
The file should be fine, as I understand from spec, there are just rules how content must be defined, but not which ones are mandatory. E.g. `[Desktop]` alone in the file should be fine. I can only tell that 
Gnome Shell and Whiskermenu hide such desktop files accordingly.

Merging: No. See it more as file overlay.

>  If multiple files have the same desktop file ID, the first one in the $XDG_DATA_DIRS precedence order is used. 

Additionaly `$XDG_DATA_HOME` which defaults to `~/.local/share/` has higher priority, so checking order would be in my case:


<table><tr><td><img src="https://user-images.githubusercontent.com/6735650/39566489-0c195cfc-4ebc-11e8-8eb5-beb259372c0f.png"></td><td>
/home/gregor/.local/share<br><br>
/usr/local/share<br><br>
/usr/share/
</td>
</tr></table>

<https://standards.freedesktop.org/desktop-entry-spec/latest/ar01s02.html>

**Comment by gmc-holle** [comment]
Strange. If they should not be merged then xfdashboard should work correctly as it uses the first one found in exactly the order you described. But the one in your `$XDG_DATA_HOME/appllications` folder is incomplete and therefore not usable. I use gaston to load and validate the file and as whisker-menu and Xfce's default application menu works for you, gaston is working correctly.

I think (but I haven't checked it) that the desktop files of all paths are merged so you can just override the default entry in your desktop file `$XDG_DATA_HOME/applications`. If so then xfdashboard has a design mistake and has a bug.

I see when I spent time to discover this issue here.

**Comment by gmc-holle** [comment]
I checked the specification at https://standards.freedesktop.org/desktop-entry-spec/latest again and I cannot find any hint that user desktop files should be merged with the system-wide desktop files . Merging means that the entries in the user desktop files override the entries at the system-wide one.

At section `Desktop File ID` is written:
`If multiple files have the same desktop file ID, the first one in the $XDG_DATA_DIRS precedence order is used.`

That the case at xfdashboard. But your desktop file is incomplete and therefore invalid and cannot be used. Looking at section `Recognized desktop entry keys` you will see we entries are missing: Type and Name (or URL if Type is `Link`). Defining only `Hidden` is - as far as I understand - invalid. And as said before I cannot find a section describing that desktop files can and should be merged.

So far I cannot see an issue with xfdashboard here.

**Comment by gsantner** [comment]
Thats exactly what I meant - no merging at all. The problem is the ID, even if it is invalid should not get scanned from /usr/share/applications then too

**Comment by gmc-holle** [comment]
Ah ok .. that is what I misunderstood. The problem with xfdashboard is that it will skip your invalid desktop file and continues searching for a valid desktop file at the remaining paths, e.g. /usr/share/applications, /usr/local/share/applications. Right? I think that can be fixed soon.

**Comment by gsantner** [comment]
Yes, now we are on the same point :D

**Comment by gmc-holle** [comment]
I had a look at the code and it is more worse than thought as the correct order the desktop file are searched and used is only correct when the application database is built initially. Any updates to any desktop files will replace the entry in the database without checking if a corresponding desktop file with higher precedence is available and unmodified, e.g. an update for an application installs a new desktop file and this file will override the user file which is still unmodifed but also has still a higher priority. So it will take more time as expected. :(

**Comment by gsantner** [comment]
I just make the Hidden=true files just with 2 more dummy fields to make them hide.
For me it's not that important anymote therefore, but compared to other applications its still a bug.

**Comment by gmc-holle** [comment]
Can you tell me these two lines you added to make it work? Or can do paste the complete file in this issue?

**Comment by gsantner** [comment]
`.local/share/applications/smplayer.desktop` (doesn't show up)
```
[Desktop Entry]
Hidden=true
Name=test
Exec=test
```

`.local/share/applications/smplayer.desktop` (shows up)
```
[Desktop Entry]
Hidden=true
```

**Comment by gsantner** [comment]
Thanks for feedback!

I would not say its an blocker for next update though, its not really critical :D.

As from other applications (like whiskermenu) I just can tell they are hidden, be it now because invalid or because Hidden is true doesnt matter in this case :D.

**Comment by gmc-holle** [comment]
I think it is fixed now in 61676fd951baa8b3b5829b91207e7b4684c90b72. Can you confirm?

**Comment by gsantner** [comment]
@gmc-holle 
Sorry thought I did answer this already :D.
Yes it works!
